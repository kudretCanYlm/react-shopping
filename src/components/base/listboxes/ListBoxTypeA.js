import React, { memo, useRef } from 'react';
import PropTypes from 'prop-types';

const ListBoxTypeA = ({ list, name, onChangeSelect = (f) => f, className }) => {
  let selected = useRef();

  const takeSelected = () => {
    onChangeSelect(selected.current.value);
  };

  return (
    <select
      ref={selected}
      onChange={takeSelected}
      name={name}
      className={`list-box list-box-type-a ${className}`}>
      {list.map((option, key) => (
        <option key={key} value={option.value}>
          {option.name}
        </option>
      ))}
    </select>
  );
};

//proptypes
ListBoxTypeA.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.any.isRequired,
      value: PropTypes.any.isRequired
    })
  ),
  name: PropTypes.string.isRequired,
  className: PropTypes.string
};

//default props
ListBoxTypeA.defaultProps = {
  list: [
    {
      name: 'empty',
      value: 'empty'
    }
  ],
  name: 'not named',
  className: ''
};

export default memo(ListBoxTypeA);
