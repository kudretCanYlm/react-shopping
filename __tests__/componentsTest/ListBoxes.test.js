import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
//componets
import ListBoxTypeA from 'components/base/listboxes/ListBoxTypeA';

Enzyme.configure({ adapter: new Adapter() });

//variables
const listBoxClassName = 'test classname';
const listBoxName = 'test name';
let selectValue = 'no selected';
const listboxList = [
  {
    name: 'name-a',
    value: 'value-a'
  },
  {
    name: 'name-b',
    value: 'value-b'
  },
  {
    name: 'name-c',
    value: 'value-c'
  },
  {
    name: 'name-d',
    value: 'value-d'
  }
];

const onChangeSelect = (value) => {
  selectValue = value;
};

describe('ListBoxTypeA Test', () => {
  it('renderer correctly', () => {
    const component = renderer
      .create(
        <ListBoxTypeA
          className={listBoxClassName}
          name={listBoxName}
          list={listboxList}
          onChangeSelect={onChangeSelect}
        />
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  it('it has my className,name and list', () => {
    const ListBoxComponent = shallow(
      <ListBoxTypeA
        className={listBoxClassName}
        name={listBoxName}
        list={listboxList}
        onChangeSelect={onChangeSelect}
      />
    );

    expect(ListBoxComponent.hasClass(listBoxClassName)).toEqual(true);
    expect(ListBoxComponent.html()).toContain(`name=\"${listBoxName}\"`);

    ListBoxComponent.children().map((option, i) => {
      expect(option.html()).toContain(`value=\"${listboxList[i].value}\"`);
      expect(option.html()).toContain(listboxList[i].name);
    });
  });

  it('onClick work', () => {
    const ListBoxComponent = shallow(
      <ListBoxTypeA
        className={listBoxClassName}
        name={listBoxName}
        list={listboxList}
        onChangeSelect={onChangeSelect}
      />
    );

    ListBoxComponent.at(0).simulate('change', {
      // target: { value: listboxList[0].value, name:listboxList[0].name }
      selectedIndex: { index: 0, text: listboxList[0].name }
    });

    expect(selectValue).toEqual('te');
  });
});
