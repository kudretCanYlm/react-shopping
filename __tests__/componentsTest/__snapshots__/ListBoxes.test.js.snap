// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBoxTypeA Test renderer correctly 1`] = `
<select
  className="list-box list-box-type-a test classname"
  name="test name"
  onChange={[Function]}
>
  <option
    value="value-a"
  >
    name-a
  </option>
  <option
    value="value-b"
  >
    name-b
  </option>
  <option
    value="value-c"
  >
    name-c
  </option>
  <option
    value="value-d"
  >
    name-d
  </option>
</select>
`;
